### DISCUSSION TITLE
Support for styling with custom fonts like

### DISCUSSION THREAD

**Discussion by dpc** [helix-editor/helix#8790] [General]
Recently published [Monaspace](https://monaspace.githubnext.com/) family of monospace introduces a cool idea of being able to use multiple different styles of the same font, with same metrics, together at the time, extending usual possibility of using italics and bold fonts

It would be great if CLI terminal emulators and applications (especially text editors) could use it as well. It could open a lot of room for innovation in the CLI UX space. Things like different elements of tmux UI, shell and other auto-completions, interactive tools, text-editing etc. could benefit from it.

It is however at the point of a chicken and egg problem. Applications don't use it because terminals don't support it, terminals don't implement it because nothing yet uses it. Font creators don't produce metric-compatible fonts, because it's not very common.

I was told that xterm does include some form of support for it already.

I was aware that Kitty would be in a good position to support it, as it's generally very featureful. The author considers it easy to implement, but [unless there's some major CLI text editor that commits to supporting it, he doesn't want to spend the effort](https://github.com/kovidgoyal/kitty/issues/6804#issuecomment-1805576628).

Well, I think Helix being so featureful CLI text editor would also be in a great position to support it, so maybe we could make it happen, get some cool feature to showcase to the users, and in the process bootstrap a nice advancement in the overall state of the art.

So the question is: would Helix be willing to add support for it, and then finding someone that is willing to do the effort.

**Comment by pascalkuthe** [comment]
I am personally not really interested in this. It's not portable since fonts need to be installed first
 This kind of styling wouls then tie a a specific theme to a specific font. There would also be yet again the issue of emulator feature detection. Without a way to reliably detect whether an emulator supports these sequences I would be opposed to adding support. I think if we want to get fancy with fonts that should just wait until we have a gui (and even then I am not sure if that's really somethingwe need to support).


It's a lot of work for something that only works with a single font (and using multiple different fonts that were not designed for it seems super no he, basically no ditor supports that).

Maybe some of the other maintainers would be interested but I am not.

**Reply by archseer** [reply]
I think it wouldn't be much work, we'd just add it as an option to the Style struct and make it configurable in the theme, the same way we do or other styling. 

(I personally also don't have any interest in this new font, out of the 5 variations only the base one looks decent)

**Reply by chtenb** [reply]
We already support ansi modifiers like dim and blink. These can be interpreted by terminal emulators to be a certain font and font style. This is how I use the font from the OP

**Comment by dpc** [comment]
[I was told that support for functionality like this](https://github.com/kovidgoyal/kitty/issues/6804#issuecomment-1893278226) landed in Vim: https://github.com/vim/vim/commit/a606f3ac036e5f3dc313f620e6b4bc00812314f9

**Comment by philocalyst** [comment]
Yeah, in light of the vim commit, I would like for this to be re-opened. This is really useful for things outside of super-families! Mixing and matching type faces has, basically since typesettings inception, been part of communicating and delineating information. The implementation would be adding a non-breaking filed to the Style struct, and some basic wiring logic.

Considering the depth of configuration a user would have to go through in order to implement this, portability isn't a major concern in my eyes; if it's communicated plainly in the docs, all would be swell. 

@pascalkuthe @archseer

**Reply by archseer** [reply]
This discussion is still open, as I said I'd be interested in supporting it but I personally don't use such a font. Someone needs to put in the effort to build this out, test it and open a PR.

**Reply by archseer** [reply]
As Pascal points out, this adds font specific theming so we couldn't offer it out of the box on themes without tying a theme to a specific font, so it would be up to the user to override their theme config as well.

**Reply by philocalyst** [reply]
> This discussion is still open, as I said I'd be interested in supporting it but I personally don't use such a font. Someone needs to put in the effort to build this out, test it and open a PR.

Just wanted to emphasize that this _isn't font specific_ -- you could mix any group of fonts.

**Reply by archseer** [reply]
You can mix any group of fonts but you have to set attributes in the theme that are specific to that font. It's possible some of the attributes overlap between different font files but there's no standard from what I can tell.
